airbyte-cdk/python/airbyte_cdk/sources/file_based/file_based_source.py

```diff
@@ -90,6 +91,7 @@ def streams(self, config: Mapping[str, Any]) -> List[Stream]:
         """
         try:
             parsed_config = self.spec_class(**config)
+            self.stream_reader.config = parsed_config
```

Comment:
why set the config as a stream_reader property instead of passing it to the methods that need it?

As it is, the stream_reader isn't fully initialized until we set its config, which is a non-obvious gotcha to keep in mind (similar to how we need to instantiate a stream's state separately from it's instantiation)

Reply:
just copying the convo over to continue here

> To give some context to other reviewers - the basic background here is that config information is needed by both the FileBasedSource and the StreamReader, but for different reasons; FileBasedSource needs it to configure streams, and the StreamReader needs it so that it knows how to, e.g. authenticate with the storage system.

In this PR, I'm allowing FileBasedSource to read the config from disk and parse it (like normal), and once parsed, the source sets the config on the StreamReader. However, the type of config accepted by FileBasedSource is AbstractFileBasedSpec, and it only cares about keys that are source-agnostic, whereas StreamReader cares about keys that are specific to the 3rd party that it's reading from. For example, S3 will be looking for a aws_access_token_id key, but FileBasedSource won't itself need that.

So that leads us to this situation, where the S3 StreamReader's config setter requires a config of type (S3) Config, but the interface config setter takes an AbstractFileBasedSpec, which doesn't have S3-specific keys. To solve this, we assert that we were given the correct Config type for our type of StreamReader.

One alternative route that I went down involved reading in the config prior to the initialization of the Source, so that it could be given to the StreamReader as an argument. Since the Source requires a StreamReader, it could get the config off of that. This is somewhat undesirable because we end up reading the config twice (because AbstractSource still reads it deep within the CDK code), and it also deviates from the pattern of letting the Source validate the config, which may have error handling behavior that we want to keep. For those two reasons I'd prefer to keep the code as-is. But I'm open to other opinions.

Thanks for adding the comment in the setter. The aspect Alex brought up makes sense, and then there's the drawback that every file based implementation will have to remember to check the instance type in their setter. But a better solution still isn't obvious to me (although I'm not sure if we would end up with a type mismatch that would still pass validation)

Reply:
We could probably remove the property setter by passing the config as a parameter to `open_file`, and `get_matching_files`. That would however force us to verify the config in both methods instead of only in the setter. 

I'm fine with leaving as-is then

Reply:
Yeah I think my preference would be to leave it as-is for now. It's a bit awkward but is not going to lead to an insidious issue - if the stream reader needs the config (which the S3 connector does) there will be a [total failure](https://github.com/airbytehq/airbyte/pull/28786/files#diff-b5e8d7cf7e79880dd2e77408410825335670decbc45f4960d2cf07ac323bae86R48) if it's not set.